zkevm-circuits/src/keccak_circuit/keccak_bit.rs

```diff
@@ -0,0 +1,870 @@
+                    + b[pi][1][k].clone()
+                    + b[pi][2][k].clone()
+                    + b[pi][3][k].clone()
+                    + b[pi][4][k].clone()
+                    + b[ni][0][pk].clone()
+                    + b[ni][1][pk].clone()
+                    + b[ni][2][pk].clone()
+                    + b[ni][3][pk].clone()
+                    + b[ni][4][pk].clone());*/
+                let bit = xor::expr(b[pi][0][k].clone(), b[pi][1][k].clone())
+                    + xor::expr(b[pi][2][k].clone(), b[pi][3][k].clone())
+                    + xor::expr(b[pi][4][k].clone(), b[ni][0][pk].clone())
+                    + xor::expr(b[ni][1][pk].clone(), b[ni][2][pk].clone())
+                    + xor::expr(b[ni][3][pk].clone(), b[ni][4][pk].clone());
+                /*input = input * MAX_INPUT_THETA_LOOKUP.expr()
+                + xor::expr(
+                    xor::expr(b[pi][0][k].clone(), b[pi][1][k].clone()),
+                    xor::expr(b[pi][2][k].clone(), b[pi][3][k].clone()),
+                )
+                + xor::expr(
+                    xor::expr(b[pi][4][k].clone(), b[ni][0][pk].clone()),
+                    xor::expr(b[ni][1][pk].clone(), b[ni][2][pk].clone()),
+                )
+                + xor::expr(b[ni][3][pk].clone(), b[ni][4][pk].clone());*/
+                theta.push((c.clone(), bit));
+            }
+        }
+
+        let mut lookup_counter = 0;
+        for chunk in theta.chunks(num_bits_per_theta_lookup) {
+            meta.lookup("Theta c", |_| {
+                let mut factor = 1u64;
+                let mut input = 0u64.expr();
+                let mut tables = Vec::new();
+                for (idx, (bit, expr)) in chunk.iter().enumerate() {
+                    //input = input * MAX_INPUT_THETA_LOOKUP.expr() + expr.clone();
+                    input = input + expr.clone() * factor.expr();
+                    factor *= MAX_INPUT_THETA_LOOKUP;
+                    tables.push((bit.clone(), c_table[1 + idx]));
+                }
+                tables.push((input, c_table[0]));
+                tables
+            });
+            lookup_counter += 1;
+        }
+        println!("Lookups: {}", lookup_counter);
+
+        meta.create_gate("boolean checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            // Absorb bits
+            for a in &a {
+                cb.require_boolean("boolean state bit", a.clone());
+            }
+
+            // q_end
+            cb.require_boolean("boolean q_end", meta.query_advice(q_end, Rotation::cur()));
+
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("first row", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            // q_end
+            cb.require_equal(
+                "q_end needs to be enabled on the first row",
+                meta.query_advice(q_end, Rotation::cur()),
+                1.expr(),
+            );
```

Comment:
Why not just use a `Fixed` or `Constant` for `q_end`?

Reply:
I guess `q_end` is a bad name. The goal is for it to be used to mark the end of one keccak hash so a new keccak hash can be started. So where this will be set to 1 will change depending on the inputs.